### PR TITLE
Add attachment support to emails

### PR DESCRIPTION
**Pull request by nickvergessen**
For testing, apply the following patch and then share something by mail
```diff
diff --git a/apps/sharebymail/lib/ShareByMailProvider.php b/apps/sharebymail/lib/ShareByMailProvider.php
index f610a1a5fa..7cab340368 100644
--- a/apps/sharebymail/lib/ShareByMailProvider.php
+++ b/apps/sharebymail/lib/ShareByMailProvider.php
@@ -428,6 +428,16 @@ class ShareByMailProvider implements IShareProvider {
                }
 
                $message->useTemplate($emailTemplate);
+               $a = $message->createAttachment();
+               $a->setFilename('abc.txt')
+                       ->setContentType('text/plain')
+                       ->setBody('abcdefghijklmnopqrstuvwxyz');
+               $message->attach($a);
+               $a = $message->createAttachment();
+               $a->setFilename('background.png')
+                       ->setContentType('image/png')
+                       ->setBody(file_get_contents(__DIR__ . '/../../../core/img/background.png'));
+               $message->attach($a);
                $this->mailer->send($message);
        }
 
```

### CODE REVIEW

**Comment by georgehrke** [comment]
Code looks good, can we have a test for `lib/private/Mail/Attachment.php`?

**Review by LukasReschke — CHANGES_REQUESTED** [review]
> ->setBody(file_get_contents(__DIR__ . '/../../../core/img/background.png'));

Your example seems to say we should also add a wrapper for Swift_Attachment::fromPath() which will also save memory 
🙂

See also https://swiftmailer.symfony.com/docs/messages.html#attaching-dynamic-content

**Comment by nickvergessen** [comment]
Done

File: lib/private/Mail/Message.php [hunk 1]
```diff
@@ -44,6 +45,38 @@ public function __construct(Swift_Message $swiftMessage) {
 	}
 
 	/**
+	 * @param string $data
+	 * @param string $filename
+	 * @param string $contentType
```
Comment (LukasReschke) [inline lib/private/Mail/Message.php]: string|null

**Review by LukasReschke — APPROVED** [review]

File: lib/private/Mail/Message.php [hunk 2]
```diff
@@ -44,6 +45,38 @@ public function __construct(Swift_Message $swiftMessage) {
...
+	public function createAttachmentFromPath($path, $contentType = null) {
+		return new Attachment(\Swift_Attachment::fromPath($path, $contentType));
+	}
+
+	/**
+	 * @param IAttachment $attachment
+	 * @return $this
+	 * @since 13.0.0
+	 */
+	public function attach(IAttachment $attachment) {
```
Comment (ChristophWurst) [inline lib/private/Mail/Message.php]: API-ergonomics-wise I'm curious why an attachment has to be added while it's created by a factory method of the message. For convenience, the attachment could be attached right away. Or am I missing a use case where you attache an attachment to another message maybe?

Comment (ChristophWurst) [reply lib/private/Mail/Message.php]: Somehow I think that the `createAttachment` method would better fit into the `IMailer` interface https://github.com/nextcloud/server/blob/master/lib/public/Mail/IMailer.php

Comment (nickvergessen) [reply lib/private/Mail/Message.php]: Done

**Review by rullzer — APPROVED** [review]
Fine by me. @ChristophWurst you are also fine now?

**Review by georgehrke — APPROVED** [review]

**Pull request merged**